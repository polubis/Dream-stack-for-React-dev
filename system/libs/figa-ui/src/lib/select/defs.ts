import type { ReactNode } from 'react';

type SelectOptionKey = string;

interface SelectOption {
  child: ReactNode;
  key: SelectOptionKey;
}

interface SelectProps {
  className?: string;
  placeholder?: ReactNode;
  initialOpen?: boolean;
  value?: SelectOptionKey;
  options: SelectOption[];
  onChange: (key: SelectOptionKey) => void;
}

export type { SelectProps, SelectOption, SelectOptionKey };
