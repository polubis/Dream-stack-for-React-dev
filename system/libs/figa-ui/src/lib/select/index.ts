export * from './select';
export * from './defs';
