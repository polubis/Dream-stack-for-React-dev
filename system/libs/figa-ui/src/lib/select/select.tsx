import type { SelectOptionKey, SelectProps } from './defs';
import type { MouseEventHandler } from 'react';

import { useMemo } from 'react';
import c from 'classnames';
import { Popover } from '../popover';

const Trigger = <K extends SelectOptionKey = SelectOptionKey>({
  className,
  value,
  options,
  placeholder,
}: Pick<SelectProps<K>, 'className' | 'value' | 'options' | 'placeholder'>) => {
  const { toggle, opened } = Popover.use();

  const valueToDisplay = useMemo(
    () =>
      value
        ? options.find((option) => option.key === value)?.child ?? placeholder
        : placeholder,
    [value, options, placeholder]
  );

  return (
    <Popover.Trigger>
      <button
        className={c('select-expander', className, {
          'select-expander-empty': value === '' || value === undefined,
          'select-expander-opened': opened,
        })}
        onClick={toggle}
      >
        {valueToDisplay}
      </button>
    </Popover.Trigger>
  );
};

const Content = <K extends SelectOptionKey = SelectOptionKey>({
  options,
  value,
  onChange,
}: Pick<SelectProps<K>, 'value' | 'options' | 'onChange'>) => {
  const { close } = Popover.use();

  const handleChange: MouseEventHandler<HTMLLIElement> = (e) => {
    const key = e.currentTarget.getAttribute('data-key');

    if (!key) {
      throw Error('Lack of key inside select list item');
    }

    close();
    onChange(key as K);
  };

  return (
    <Popover.Content fullWidth>
      <ul className="select-list">
        {options.map(({ key, child }) => (
          <li
            key={key}
            data-key={key}
            className={c('select-list-option', {
              'select-list-option-active': key === value,
            })}
            onClick={handleChange}
          >
            {child}
          </li>
        ))}
      </ul>
    </Popover.Content>
  );
};

const Select = <K extends SelectOptionKey = SelectOptionKey>({
  className,
  placeholder = 'Choose an option',
  value,
  options,
  initialOpen,
  onChange,
}: SelectProps<K>) => {
  return (
    <Popover openOnInit={initialOpen} closeMode="backdrop">
      <Trigger
        className={className}
        value={value}
        options={options}
        placeholder={placeholder}
      />
      <Content options={options} onChange={onChange} value={value} />
    </Popover>
  );
};

export { Select };
