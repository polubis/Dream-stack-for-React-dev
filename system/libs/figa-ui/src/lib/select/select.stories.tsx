import type { Story, Meta } from '@storybook/react';

import { Select } from './select';
import { useState } from 'react';

export default {
  component: Select,
  title: 'Select',
} as Meta;

const Template: Story = () => {
  const [value, setValue] = useState('');

  const options = [
    {
      key: '0',
      child: <div>1 child</div>,
    },
    {
      key: '1',
      child: <div>2 child</div>,
    },
    {
      key: '2',
      child: <div>3 child</div>,
    },
    {
      key: '3',
      child: <div>4 child</div>,
    },
  ];

  return (
    <div
      style={{
        display: 'flex',
        flexFlow: 'wrap',
        gap: '20px',
      }}
    >
      <Select
        options={options}
        value={value}
        placeholder="My custom placeholder"
        onChange={setValue}
      />
      <div>
        <Select
          options={options}
          initialOpen
          value={value}
          onChange={setValue}
        />
      </div>
      <div>
        <Select options={options} value={value} onChange={setValue} />
      </div>
      <div>
        <Select options={options} value={options[0].key} onChange={setValue} />
      </div>
      <div>
        <Select
          options={options}
          initialOpen
          value={options[0].key}
          onChange={setValue}
        />
      </div>
      <Select
        options={options}
        value={value}
        placeholder="My custom placeholder"
        onChange={setValue}
      />
    </div>
  );
};

export const Default = Template.bind({});
Default.args = {};
