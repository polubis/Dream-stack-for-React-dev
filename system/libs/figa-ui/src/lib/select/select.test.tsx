import type { SelectOption } from './defs';

import { render, fireEvent, screen } from '@testing-library/react';
import { useState } from 'react';
import { Select } from './select';

describe('select can be used when: ', () => {
  const OPTIONS: SelectOption[] = [
    {
      key: '0',
      child: <div>Child 0</div>,
    },
    {
      key: '1',
      child: <div>Child 1</div>,
    },
  ];

  it('[FRAGILE] renders with default props', () => {
    const { asFragment } = render(<Select options={[]} onChange={jest.fn()} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('allows to pass custom placeholder', () => {
    render(
      <Select placeholder="placeholder" options={[]} onChange={jest.fn()} />
    );

    screen.getByText(/placeholder/);
  });

  it('options are hidden by default', () => {
    render(<Select options={OPTIONS} onChange={jest.fn()} />);

    expect(screen.queryByText(/Child 0/)).toBeFalsy();
    expect(screen.queryByText(/Child 1/)).toBeFalsy();
  });

  it('displays options initially', () => {
    render(<Select options={OPTIONS} initialOpen onChange={jest.fn()} />);

    screen.getByText(/Child 0/);
    screen.getByText(/Child 1/);
  });

  it('opens option menu after click', () => {
    const Fixture = () => {
      const [value, setValue] = useState('');

      return (
        <Select
          value={value}
          options={OPTIONS}
          placeholder="Trigger"
          onChange={setValue}
        />
      );
    };

    render(<Fixture />);

    fireEvent.click(screen.getByText(/Trigger/));

    screen.getByText(/Child 0/);
    screen.getByText(/Child 1/);
  });

  it('closes menu after element choose', () => {
    const Fixture = () => {
      const [value, setValue] = useState('');

      return (
        <Select
          value={value}
          options={OPTIONS}
          placeholder="Trigger"
          onChange={setValue}
        />
      );
    };

    render(<Fixture />);

    fireEvent.click(screen.getByText(/Trigger/));

    screen.getByText(/Child 0/);
    screen.getByText(/Child 1/);

    fireEvent.click(screen.getByText(/Child 0/));

    screen.getByText(/Child 0/);
    expect(screen.queryByText(/Child 1/)).toBeFalsy();
  });

  it('closes menu after outside click', () => {
    const Fixture = () => {
      const [value, setValue] = useState('');

      return (
        <>
          <button>Click outside</button>

          <Select
            value={value}
            options={OPTIONS}
            placeholder="Trigger"
            onChange={setValue}
          />
        </>
      );
    };

    render(<Fixture />);

    fireEvent.click(screen.getByText(/Trigger/));

    screen.getByText(/Child 0/);
    screen.getByText(/Child 1/);

    fireEvent.click(screen.getByText(/Click outside/));

    screen.getByText(/Trigger/);
  });

  describe('[FRAGILE] assigns correct HTML metadata when', () => {
    it('opened', () => {
      const { asFragment } = render(
        <Select options={OPTIONS} initialOpen onChange={jest.fn()} />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('closed', () => {
      const { asFragment } = render(
        <Select options={OPTIONS} onChange={jest.fn()} />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('element is choosed', () => {
      const { asFragment } = render(
        <Select options={OPTIONS} placeholder="Trigger" onChange={jest.fn()} />
      );

      fireEvent.click(screen.getByText(/Trigger/));

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
