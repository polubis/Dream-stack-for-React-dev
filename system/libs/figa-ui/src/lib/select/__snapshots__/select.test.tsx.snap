// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select can be used when:  [FRAGILE] assigns correct HTML metadata when closed 1`] = `
<DocumentFragment>
  <div
    class="select"
  >
    <button
      class="select-expander select-expander-empty"
    >
      Choose an option
    </button>
  </div>
</DocumentFragment>
`;

exports[`select can be used when:  [FRAGILE] assigns correct HTML metadata when element is choosed 1`] = `
<DocumentFragment>
  <div
    class="select"
  >
    <button
      class="select-expander select-expander-empty select-expander-opened"
    >
      Trigger
    </button>
    <ul
      class="select-list"
    >
      <li
        class="select-list-option"
        data-key="0"
      >
        <div>
          Child 0
        </div>
      </li>
      <li
        class="select-list-option"
        data-key="1"
      >
        <div>
          Child 1
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`select can be used when:  [FRAGILE] assigns correct HTML metadata when opened 1`] = `
<DocumentFragment>
  <div
    class="select"
  >
    <button
      class="select-expander select-expander-empty select-expander-opened"
    >
      Choose an option
    </button>
    <ul
      class="select-list"
    >
      <li
        class="select-list-option"
        data-key="0"
      >
        <div>
          Child 0
        </div>
      </li>
      <li
        class="select-list-option"
        data-key="1"
      >
        <div>
          Child 1
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`select can be used when:  renders with default props 1`] = `
<DocumentFragment>
  <div
    class="select"
  >
    <button
      class="select-expander select-expander-empty"
    >
      Choose an option
    </button>
  </div>
</DocumentFragment>
`;
